cumulus/client/cirrus-executor/src/lib.rs

```diff
@@ -388,22 +399,78 @@ where
 	/// Checks the execution receipt from the executor peers.
 	fn on_execution_receipt(
 		&self,
-		_execution_receipt: &ExecutionReceipt<<Block as BlockT>::Hash>,
+		execution_receipt: &ExecutionReceipt<<Block as BlockT>::Hash>,
 	) -> Result<Action, Self::Error> {
 		// TODO: validate the Proof-of-Election
 
-		// TODO: check if the received ER is same with the one produced locally.
-		let same_with_produced_locally = true;
-
-		if same_with_produced_locally {
-			Ok(Action::RebroadcastExecutionReceipt)
+		// TODO: more efficient execution receipt checking strategy?
+		let local_receipt = if let Some(local_receipt) =
+			crate::aux_schema::load_execution_receipt::<_, Block>(
+				&*self.client,
+				execution_receipt.secondary_hash,
+			)? {
+			local_receipt
 		} else {
+			// Wait for the local execution receipt until it's ready.
+			let (tx, rx) = crossbeam::channel::bounded::<
+				sp_blockchain::Result<ExecutionReceipt<Block::Hash>>,
+			>(1);
+			let client = self.client.clone();
+			let block_hash = execution_receipt.secondary_hash;
+			self.spawner.spawn(
+				"wait-for-local-execution-receipt",
+				None,
+				async move {
+					loop {
+						match crate::aux_schema::load_execution_receipt::<_, Block>(
+							&*client, block_hash,
+						) {
+							Ok(Some(local_receipt)) => {
+								let _ = tx.send(Ok(local_receipt));
+								break
+							},
+							Ok(None) => {
+								tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+							},
+							Err(e) => {
+								let _ = tx.send(Err(e));
+								break
+							},
+						}
+					}
+				}
+				.boxed(),
+			);
+
+			rx.recv()??
+		};
+
+		// TODO: What happens for this obvious error?
```

Comment:
Same as in any other case, you compare trace item one by one and generate fraud proof for item that is wrong or missing

Reply:
There is a possible case that an external ER that contains all the valid roots and some extra ones? In that case comparing on the iterator won't detect it, but I believe the block will still fail in the verification somewhere else.

Reply:
Everything is possible, adversary can create whatever ER they want :slightly_smiling_face: